### PR TITLE
Support filling forecast gaps

### LINKED ISSUE #459: Add CLI command to backfill observations/forecasts
Including NWP and persistence forecasts. This likely searches an observation/forecast series for missing chunks and calls the appropriate functions with these periods to fill in the data.

### PR DESCRIPTION
**Pull request by alorenzo175**
- [x] Closes #459.
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

File: solarforecastarbiter/reference_forecasts/main.py [hunk 1]
```diff
@@ -417,38 +417,42 @@ def process_nwp_forecast_groups(session, run_time, forecast_df):
...
+    errors = _verify_nwp_forecasts_compatible(group)
+    if errors:
+        logger.error(
+            'Not all forecasts compatible in group with %s. '
+            'The following parameters may differ: %s', run_for, errors)
+        return
+    try:
+        key_fx = group.loc[run_for].forecast
+    except KeyError:
+        logger.error('Forecast, %s,  that others are piggybacking on not '
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/main.py:434]:
```suggestion
        logger.error('Forecast, %s, that others are piggybacking on not '
```

File: solarforecastarbiter/reference_forecasts/main.py [hunk 2]
```diff
@@ -473,19 +477,75 @@ def make_latest_nwp_forecasts(token, run_time, issue_buffer, base_url=None):
...
+    next_issue_time = utils.find_next_issue_time_from_last_forecast(
+            fx, last_timestamp)
+    while next_issue_time < max_run_time:
+        yield next_issue_time
+        next_issue_time = utils.get_next_issue_time(
+            fx, next_issue_time + pd.Timedelta('1ns'))
+
+
+def _find_group_gaps(session, forecasts, start, end):
+    times = set()
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/main.py:514]: if you can trust the order from the generator, which I think you can, you can use `dict.fromkeys` as done here: https://github.com/SolarArbiter/solarforecastarbiter-core/blob/839f4ecc61b52fa43fedc29c3020f0339eac78e9/solarforecastarbiter/metrics/preprocessing.py#L293-L298 Not clear to me that one is more readable than the other. I'm sure performance doesn't matter for this.

File: solarforecastarbiter/reference_forecasts/main.py [hunk 3]
```diff
@@ -473,19 +477,75 @@ def make_latest_nwp_forecasts(token, run_time, issue_buffer, base_url=None):
...
-    execute_for = forecast_df[
-        forecast_df.next_issue_time <= run_time + issue_buffer]
-    if execute_for.empty:
-        logger.info('No forecasts to be made at %s', run_time)
-        return
-    process_nwp_forecast_groups(session, run_time, execute_for)
+    return session, forecast_df
+
+
+def _nwp_issue_time_generator(fx, last_timestamp, next_timestamp):
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/main.py:500]: I feel like I understand what this function is doing and how it's used in the calling loop, but I don't understand the parameter names. Nor do I understand the comments. So I'm not confident I really know what's going on.

I thought the parameter names would be `gap_start` and `gap_end`. I guess that's similar to `last_existing_timestamp` and `next_existing_timestamp` and if you've taken out "existing" for brevity then we're on the same page.

File: solarforecastarbiter/reference_forecasts/main.py [hunk 4]
```diff
@@ -654,24 +724,40 @@ def make_latest_persistence_forecasts(token, max_run_time, base_url=None):
...
         if len(serlist) > 0:
             ser = pd.concat(serlist)
-            for cser in generate_continuous_chunks(ser, fx.interval_length):
-                session.post_forecast_values(fx.forecast_id, cser)
+            for cser in generate_continuous_chunks(
+                    ser, fx.interval_length):
+                if type(fx) == datamodel.Forecast:
+                    session.post_forecast_values(id_, cser)
+                else:
+                    session.post_probabilistic_forecast_constant_value_values(  # NOQA
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/main.py:759]: what's noqa for? looks like it's less than 80 characters on my diff

File: solarforecastarbiter/reference_forecasts/main.py [hunk 5]
```diff
@@ -654,24 +724,40 @@ def make_latest_persistence_forecasts(token, max_run_time, base_url=None):
...
             else:
-                serlist.append(fx_ser)
+                out[fx.forecast_id].append(fx_out)
+    for id_, serlist in out.items():
         if len(serlist) > 0:
             ser = pd.concat(serlist)
-            for cser in generate_continuous_chunks(ser, fx.interval_length):
-                session.post_forecast_values(fx.forecast_id, cser)
+            for cser in generate_continuous_chunks(
+                    ser, fx.interval_length):
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/main.py:755]: looks to me like this still fits on 1 line

File: solarforecastarbiter/reference_forecasts/main.py [hunk 6]
```diff
@@ -694,28 +780,115 @@ def make_latest_probabilistic_persistence_forecasts(
...
+    start : pandas.Timestamp
+        The start of the period to search for missing forecast values.
+    end : pandas.Timestamp
+        The end of the period to search for missing forecast values.
+
+    Returns
+    -------
+    generator of (Forecast, Observation, index, data_start, data_end, issue_times)
+
+    """  # NOQA
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/main.py:816]:
```suggestion
    """  # NOQA: E501
```

File: solarforecastarbiter/reference_forecasts/main.py [hunk 7]
```diff
@@ -694,28 +780,115 @@ def make_latest_probabilistic_persistence_forecasts(
...
+        issue_times = tuple(sorted(times))
+        if len(issue_times) == 0:
+            continue
+
+        data_start, data_end = utils.get_data_start_end(
+            observation, fx, issue_times[0], issue_times[-1])
+        out = namedtuple(
+            'PersistenceGapParameters',
+            ['forecast', 'observation', 'index', 'data_start', 'data_end',
+             'issue_times'])
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/main.py:841]: define above loop?

File: solarforecastarbiter/reference_forecasts/main.py [hunk 8]
```diff
@@ -694,28 +780,115 @@ def make_latest_probabilistic_persistence_forecasts(
...
+    observations = session.list_observations()
+    params = generate_reference_persistence_forecast_gaps_parameters(
+        session, forecasts, observations, start, end)
+    for fx, obs, index, data_start, data_end, issue_times in params:
+        _pers_loop(session, fx, obs, index, data_start, data_end,
+                   issue_times)
+
+
+def fill_persistence_forecasts_gaps(
+        token, start, end, base_url=None):
```
Comment (wholmgren) [inline solarforecastarbiter/reference_forecasts/main.py:857]:
```suggestion
def fill_persistence_forecasts_gaps(token, start, end, base_url=None):
```

**Review by wholmgren — APPROVED** [review]

File: solarforecastarbiter/reference_forecasts/main.py (hunk 3 above)
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/main.py:500]: yes, took out "existing". modified the parameters and expanded comments a bit

File: solarforecastarbiter/reference_forecasts/main.py (hunk 2 above)
Comment (alorenzo175) [reply solarforecastarbiter/reference_forecasts/main.py:514]: for a single forecast, I think that would be fine. But for multiple forecasts that may have different gaps, I think sticking with the explicit sort is safer and possibly required

File: solarforecastarbiter/io/api.py [hunk 9]
```diff
@@ -864,6 +921,36 @@ def get_probabilistic_forecast_constant_value_values(
         return adjust_timeseries_for_interval_label(
             out, interval_label, start, end)
 
+    @ensure_timestamps('start', 'end')
+    def get_probabilistic_forecast_value_gaps(
+            self, forecast_id, start, end):
+        """Get any gaps in forecast data from start to end. In addition
+        to querying the /forecasts/cdf/{forecast_id}/values/gaps
+        endpoint, this function also queries the forecast timerange of the
+        first constant value only to return all gaps from start to end.
```
[the remark this answers is not in the document]
Comment (alorenzo175) [reply solarforecastarbiter/io/api.py:930]: maybe those could use some revamping, but they mostly follow the convention that "... Probabilistic Forecast" refers to one constant value and "... Probabilistic Forecast Group" refers to the collection

**Pull request merged**